Support for MAC values, support for ISO8601 timestamps, support for string concatenation and support for underscore in identifiers

takase1121 opened the pull request:
Sorry for the long name, but the PR is indeed as explained by the title:
* Add support for MAC values
* Add support for string concatenation via producer `@cat`
* Add support for ISO8601 timestamps, Eg. `"2020-04-05T12:35:08.658Z"`
* Add support for underscores, `_` in identifier names

This PR includes #7 and my own commits.

raviranjanjha commented:
Hi,
These changes look good. Even I was looking to add the string concatenate producer.
Just one request, can you include '.' as this is also common in json identifiers.
So in antlr file, it would need to be:
IDENTIFIER : [a-zA-Z0-9_'.']+;

Do you have necessary rights for committing? If not can you contact jsontemplate2019@gmail.com <jsontemplate2019@gmail.com> including me (raviranjanjha@gmail.com).
Thanks,
Ravi

haihanyin commented:
Hi, takase1121. Really appreciate your contribution! I have gone through your changes. They look fine to me. Just one thing. The .iml file should not be included. Can you remove that from the pull request?

takase1121 commented:
Sorry for the late reply.

For @raviranjanjha,
I added the thing I requested.

For @haihanyin,
I removed the iml file. However, this file seems to be automatically generated by intelliJ IDEA so I added it to .gitignore too. Hope that would be sufficient.

And another question regarding the spec:
Why are JSON string like `"hey"` is not allowed for object key? That is kinda weird.

haihanyin commented:
Can you give me an example regarding "hey"

raviranjanjha commented:
Hi Haihanyin,
Can you kindly publish a new version with these changes? I am waiting for
them for some time :)
Kind Regards,
Ravi Ranjan Jha


On Thu, Apr 9, 2020 at 4:41 PM haihanyin <notifications@github.com> wrote:

> Can you give me an example regarding "hey"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/json-template/JsonTemplate/pull/9#issuecomment-611471394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWXHAYR3VZCTBLUEIBKSODRLWUNNANCNFSM4MAYYFWA>
> .
>

takase1121 commented:
Since JsonTemplate is kinda superset of JSON, it should be able to accept:
```json
{
    "hey": "John"
}
```
As that is valid JSON. However, the ANTLR rules only allow `IDENTIFIER` to be the key, so it will only allow this
```js
{
    hey: "John"
}
```

haihanyin commented:
> Hi Haihanyin, Can you kindly publish a new version with these changes? I am waiting for them for some time :) Kind Regards, Ravi Ranjan Jha
> […](#)
> On Thu, Apr 9, 2020 at 4:41 PM haihanyin ***@***.***> wrote: Can you give me an example regarding "hey" — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#9 (comment)](https://github.com/json-template/JsonTemplate/pull/9#issuecomment-611471394)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAWXHAYR3VZCTBLUEIBKSODRLWUNNANCNFSM4MAYYFWA> .

The new version 0.2.2 has been deployed.

haihanyin commented:
The new version 0.2.2 has been deployed.

haihanyin commented:
@takase1121 
1. Json always takes strings as keys. There is no need to write quotes in JsonTemplate. JsonTemplate wants to avoid quotes as much as possible.
2. The purpose of JsonTemplate is not replacing Json or being a superset of Json. It is a tool for generating Json.

takase1121 commented:
Then wouldn't it be better to at least allow computated object keys, eg: `{ @raw("help:me"): @s }`, with this at least you can insert raw JSON value if you ever came across object keys that has characters used by JSON standard, eg: `:`
